openedx/core/djangoapps/course_groups/models.py

```diff
@@ -58,6 +60,96 @@ def create(cls, name, course_id, group_type=COHORT):
         )
 
 
+class CohortMembership(models.Model):
+    """Used internally to enforce our particular definition of uniqueness"""
+
+    course_user_group = models.ForeignKey(CourseUserGroup)
+    user = models.ForeignKey(User)
+    course_id = CourseKeyField(max_length=255)
+
+    previous_cohort = None
+    previous_cohort_name = None
+    previous_cohort_id = None
+
+    class Meta(object):
+        unique_together = (('user', 'course_id'), )
+
+    # The sole purpose of overriding this method is to get the django 1.6 behavior of allowing 'validate_unique'
+    # For django 1.8 upgrade, just remove this method and allow the base method to be called instead.
+    # Reference: https://docs.djangoproject.com/en/1.6/ref/models/instances/, under "Validating Objects"
+    def full_clean(self, **kwargs):
+        self.clean_fields()
+        self.clean()
+        if 'validate_unique' not in kwargs or kwargs['validate_unique'] is True:
+            self.validate_unique()
+
+    def clean_fields(self, *args, **kwargs):
+        if self.course_id is None:
+            self.course_id = self.course_user_group.course_id
+        super(CohortMembership, self).clean_fields(*args, **kwargs)
+
+    def clean(self):
+        if self.course_user_group.group_type != CourseUserGroup.COHORT:  # pylint: disable=E1101
+            raise ValidationError("CohortMembership cannot be used with CourseGroup types other than COHORT")
+        if self.course_user_group.course_id != self.course_id:
+            raise ValidationError("Non-matching course_ids provided")
+
+    def save(self, *args, **kwargs):
+        self.full_clean(validate_unique=False)
+
+        # This loop has been created to allow for optimistic locking, and retrial in case of losing a race condition.
+        # The specific limit of 5 was chosen arbitrarily as a good number of times to attempt saving before giving up.
+        max_retries = 5
+        success = False
+        for num_attempts in range(max_retries):
+            saved_membership = None
+
+            # The following 2 "transaction" lines force a fresh read, they can be removed once we're on django 1.8
+            # http://stackoverflow.com/questions/3346124/how-do-i-force-django-to-ignore-any-caches-and-reload-data
+            with transaction.commit_manually():
+                transaction.commit()
+
+            try:
+                with transaction.commit_on_success():
+                    self.course_user_group.users.add(self.user)  # pylint: disable=E1101
+
+                    saved_membership = CohortMembership.objects.select_for_update().get(
+                        user__id=self.user.id,  # pylint: disable=E1101
+                        course_id=self.course_id
+                    )
```

Comment:
@doctoryes, how's this new approach look? It seems to handle race conditions correctly (tested using the `sleep()` calls you see here), and it looks simpler, to me at least.